src/spectroscopy_bluesky/i20/ion_autofill.py

```diff
@@ -0,0 +1,139 @@
+import enum
+
+import bluesk.plan_stubs as bps
+from dodal.common.types import MsgGenerator
+
+from spectroscopy_bluesky.i20.devices.gas_rig import GasInjectionController
+
+# todo are those parameters? those should be generic
+ionchamber2fill = "I0"  # I0, I1, iT, iRef
+targetPressureAr = 35  # mbar
+targetPressureHe = 1800  # mbar
+
+
+class GasToInject(enum.Enum):
+    ARGON = "argon"
+    HELIUM = "helium"
+    KRYPTON = "krypton"
+    NITROGEN = "nitrogen"
+
+
+class IonChamberToFill(enum.Enum):
+    I0 = "I0"
+    I1 = "I1"
+    iT = "iT"
+    iRef = "iRef"
+
+
+def get_gas_valve(gir, gas_to_inject: GasToInject):
+    gas_map = {
+        GasToInject.ARGON: gir.ar_valve,
+        GasToInject.HELIUM: gir.he_valve,
+        GasToInject.KRYPTON: gir.kr_valve,
+        GasToInject.NITROGEN: gir.n2_valve,
+    }
+    return gas_map[gas_to_inject]
+
+
+ionchamber_purge_time = 20.00  # 30
+ionchamber_leak_wait_time = 10.0  # 10
+injection_equilibration_wait_time = 20
+helium_equilibration_wait_time = 10.0
+
+
+def get_chamber_valve(gir, ion_chamber: IonChamberToFill):
+    # Map chamber to the correct valve only
+    chamber_valve_map = {
+        IonChamberToFill.I0: gir.i0_valve,
+        IonChamberToFill.I1: gir.i1_valve,
+        IonChamberToFill.iT: gir.it_valve,
+        IonChamberToFill.iRef: gir.iref_valve,
+    }
+    return chamber_valve_map[ion_chamber]
+
+
+def ion_autofill(
+    gir: GasInjectionController,
+    target_pressure_mbar: float,
+    ion_chamber: IonChamberToFill = IonChamberToFill.I0,
+    gas_to_inject: GasToInject = GasToInject.ARGON,
+) -> MsgGenerator:
+    """
+    NOTE scientifically we do small portion of the heavy gas to measure flux
+    we use the light gas (Helium) to make sure that the pressure is positive so that
+    we do not get air leak into the ion chamber.
+    """
+    chamber_valve = get_chamber_valve(gir, ion_chamber)
+    gas_valve = get_gas_valve(gir, gas_to_inject)
+    chamber_pressure = (
+        gir.pressure_controller_2
+    )  # Always use controller 2 for chamber valves
+
+    # todo when to run this?
+    def purge_line(valve, pressure: float) -> MsgGenerator:
+        # turn on the the vacuum pump - reset, open
+        yield from bps.mv(gir.vacuum_pump, 0)
+
+        # open valve line valve
+        yield from bps.mv(gir.line_valve, 2)  # reset
+        yield from bps.mv(gir.line_valve, 0)
+        # open valve
+        line_pressure = (yield from bps.read(gir.pressure_controller_1.readout))[
+            "value"
+        ]
```

Comment:
Should: I think you can just use `rd` here instead? 
```suggestion
        line_pressure = yield from bps.rd(gir.pressure_controller_1.readout)
```

Reply:
maybe. `rd` being different from `read` is an bluesky DX mishap that's why I misremembered.

Will update